test/test_tensordict.py

```diff
@@ -4271,9 +4271,54 @@ def nested_lazy_het_td(batch_size):
             {"lazy": td_stack, "dense": torch.zeros(3, 3, 2)},
             [],
         )
-        obs = obs.expand(batch_size)
+        obs = obs.expand(batch_size).clone()
         return obs
 
+    @pytest.mark.parametrize("batch_size", [(), (2,), (1, 2)])
+    @pytest.mark.parametrize("cat_dim", [0, 1, 2])
+    def test_cat_lazy_stack(self, batch_size, cat_dim):
+        if cat_dim > len(batch_size):
+            return
+        td_lazy = self.nested_lazy_het_td(batch_size)["lazy"]
+
+        res = torch.cat([td_lazy], dim=cat_dim)
+        assert assert_allclose_td(res, td_lazy)
+        assert res is not td_lazy
+        td_lazy_clone = td_lazy.clone()
+        res = torch.cat([td_lazy_clone], dim=cat_dim, out=td_lazy)
+        assert res is td_lazy
+        assert assert_allclose_td(res, td_lazy_clone)
+
+        td_lazy_2 = td_lazy.clone()
+        td_lazy_2.apply_(lambda x: x + 1)
+
+        res = torch.cat([td_lazy, td_lazy_2], dim=cat_dim)
+        assert res.stack_dim == len(batch_size)
+        assert res.shape[cat_dim] == td_lazy.shape[cat_dim] + td_lazy_2.shape[cat_dim]
+        index = (slice(None),) * cat_dim + (slice(0, td_lazy.shape[cat_dim]),)
+        assert assert_allclose_td(res[index], td_lazy)
+        index = (slice(None),) * cat_dim + (slice(td_lazy.shape[cat_dim], None),)
+        assert assert_allclose_td(res[index], td_lazy_2)
+
+        res = torch.cat([td_lazy, td_lazy_2], dim=cat_dim)
+        assert res.stack_dim == len(batch_size)
+        assert res.shape[cat_dim] == td_lazy.shape[cat_dim] + td_lazy_2.shape[cat_dim]
+        index = (slice(None),) * cat_dim + (slice(0, td_lazy.shape[cat_dim]),)
+        assert assert_allclose_td(res[index], td_lazy)
+        index = (slice(None),) * cat_dim + (slice(td_lazy.shape[cat_dim], None),)
+        assert assert_allclose_td(res[index], td_lazy_2)
+
+        if cat_dim != len(batch_size):  # cat dim is not stack dim
+            batch_size = list(batch_size)
+            batch_size[cat_dim] *= 2
+            td_lazy_dest = self.nested_lazy_het_td(batch_size)["lazy"]
+            res = torch.cat([td_lazy, td_lazy_2], dim=cat_dim, out=td_lazy_dest)
```

Comment:
when testing with `out` we should test that all the leaves of `out` before stack have the same data_ptr() of the leaves after stack:
```python
def decompose(td):
    if isinstance(td, LazyStackedTensorDict):
        for td in td.tensordicts:
            yield from decompose(td)
    else:
        for v in self.values():
            if is_tensor_collection(v):
                yield from decompose(v)
            else:
                yield v

data_ptr_set_before = {val.data_ptr() for val in decompose(out)}
torch.stack(..., out=out)
data_ptr_set_after = {val.data_ptr() for val in decompose(out)}
assert data_ptr_set_after == data_ptr_set_before
```

Reply:
well this was a great suggestion as it led to finidng bugs in the update functions which did not recursively pass args

Reply:
i added it to all the tests with out=smth so that we can sleep at night